[REG v2] Can't have integration tests with same module name

leandro-lucarella-sociomantic opened the pull request:
```
/docker/turtle/build/devel/tmp/allunittests.d(29): Error: module main from file test/dmqfull/main.d conflicts with another module main from file test/dhtrestart/main.d
```

I think this must be consequence of 145edfcd5b050736176a1d0c1436e8c47796cf34 - all those module are now built in one go and thus must have distinct module names (== explicit module statements in this turtle case).

leandro-lucarella-sociomantic commented:
So the issue here is neither [dmqfull/main.d](https://github.com/sociomantic-tsunami/dmqproto/blob/v13.x.x/test/dmqfull/main.d) nor [dhtrestart/main.d](https://github.com/sociomantic-tsunami/dhtproto/blob/v13.x.x/test/dhtrestart/main.d) have a `module` declaration, right?

I don't see how MakD could deal with this, I think you just have to have `module` declarations in you tests. But maybe we need to add it to the release notes.

mihails-strasuns-sociomantic commented:
Yep, so their module name defaults to just unqualified file name, which is `main`. Awkward thing is that I have been removing module declarations so that things can compile with 2.071.2 - and now it hits back in another away.

Fortunately Makd v2 will have `integrationtest` support as alternative way to fix 2.071.2 issue, so just explaining all this mess in release notes should be OK.

leandro-lucarella-sociomantic commented:
Could you write a migration note about this since you have the interactions more clear than me. I think I just reacted applying patches to workaround the problems you found :D

mihails-strasuns-sociomantic commented:
```
Starting with v2.0.0 Makd will compile all modules from integration tests as part of regular unittest target and stop supplying `-unittest` flag when compiling actual integration tests. This may break compilation of a project so that manual adjustment is required after Makd upgrade:

1. If integration test modules don't have module statements or have one with identical module names, it will result in compilation failure. Solution is to add explicit module statements to give each file distinct module name.
2. Integration test modules containing `main` functions need to be modified to version out that `main` function via `version(UnitTest)` or unittest target will fail to link because of multiple `main`s. 
```

leandro-lucarella-sociomantic committed:
relnotes: Improve migration steps for tests

Fixes #91.

leandro-lucarella-sociomantic commented:
Fix attached.

leandro-lucarella-sociomantic committed:
relnotes: Unify packaging notes

leandro-lucarella-sociomantic commented:
Added an unrelated improvement to migration notes (unify packaging changes).

mihails-strasuns-sociomantic reviewed (commented):
@gavin-norman-sociomantic please merge after english check

gavin-norman-sociomantic commented on relnotes/integrationtests.migration.md:
```diff
@@ -0,0 +1,16 @@
+* Unit and integration tests changes
+
+  * Starting with v2.0.0 Makd will compile all modules from integration tests as part of regular unittest target and stop supplying `-unittest` flag when compiling actual integration tests. This may break compilation of a project so some manual tweaks are required:
```
"part of the regular"

gavin-norman-sociomantic replied:
"stop supplying the `-unittest` flag"

gavin-norman-sociomantic commented on relnotes/integrationtests.migration.md:
```diff
@@ -0,0 +1,16 @@
+
+  * Starting with v2.0.0 Makd will compile all modules from integration tests as part of regular unittest target and stop supplying `-unittest` flag when compiling actual integration tests. This may break compilation of a project so some manual tweaks are required:
+
+    1. If integration test modules don't have module statements or have one with identical module names, it will result in compilation failure.
```
"or have one with identical module names" -- does this mean "or multiple modules have the same name" ? (The latter is clearer.)

leandro-lucarella-sociomantic replied:
Yes, it seems more clear

gavin-norman-sociomantic commented on relnotes/integrationtests.migration.md:
```diff
@@ -0,0 +1,16 @@
+
+    2. As with modules containing a `main()` in the regular user code in `$(SRC)`, integration test modules containing `main()` functions need to be modified to version out that `main()` function via `version(UnitTest)`, otherwise the unittest target will fail to link because of multiple `main()`s.
+
+       Again, you could manually add these modules to `TEST_FILTER_OUT`, but it is not recommended as any unit tests present in the module with the `main()` function won't be ran this way.
```
"won't be run"

gavin-norman-sociomantic commented on relnotes/integrationtests.migration.md:
```diff
@@ -0,0 +1,16 @@
+
+  * Integration tests default location changed from `test` to `integrationtest`
+
+    The location can be set explicitly via the variable `INTEGRATIONTEST`, but you are not required to define it anymore (as it was required in v1.x.x), you just need to define it if you don't want the new default (`integrationtest`). This change was introduced to avoid symbol clashes in D2.
```
"This change was introduced to avoid symbol clashes in D2." -- maybe explain this a little bit more? (Just from that sentence, I don't understand the reasoning.)

leandro-lucarella-sociomantic replied:
Added the classical example, @mihails-strasuns-sociomantic please check if what I'm saying is correct, as I might be forgetting or misinterpreting some detail.

mihails-strasuns-sociomantic replied:
I would rephrase it more to the point "can't directly use any symbol named `test` if you root package name is also `test`".

gavin-norman-sociomantic reviewed (commented):
A few small points.

leandro-lucarella-sociomantic committed:
relnotes: Mention more potential problems

Mention some more potential problems with removing `-version=UnitTest
-debug=UnitTest` when compiling integration tests.

Fixes #95.

leandro-lucarella-sociomantic commented:
Added one more clarification. Ping?

leandro-lucarella-sociomantic commented:
Updated to do all the squashing.

joseph-wakeling-sociomantic commented on relnotes/integrationtests.migration.md:
```diff
@@ -0,0 +1,17 @@
+
+    1. If integration test modules don't have module statements or multiple modules have the same name.
+
+       The solution is to add explicit module statements to give each file distinct module name.
```
`a distinct module name` ... ?

joseph-wakeling-sociomantic commented on relnotes/integrationtests.migration.md:
```diff
@@ -0,0 +1,17 @@
+
+       The solution is to add explicit module statements to give each file distinct module name.
+
+    2. As with modules containing a `main()` in the regular user code in `$(SRC)`, integration test modules containing `main()` functions need to be modified to version out that `main()` function via `version(UnitTest)`, otherwise the unittest target will fail to link because of multiple `main()`s.
```
Might be better to be entirely explicit: `version (UnitTest) {} else` ... ?

leandro-lucarella-sociomantic replied:
This is already described verbosely in another note, I think it will be too repetitive to include it again. But I can refer to the other note.

joseph-wakeling-sociomantic commented on relnotes/pkg.migration.md:
```diff
@@ -0,0 +1,19 @@
+
+    Change `from defaults import OPTS, ARGS` to `import defaults`.
+
+    This change helps making much compact utility functions. See changes on `FUN.desc()` for an example.
```
`much compact` doesn't quite make sense as phrasing -- did you mean `much more compact`?

joseph-wakeling-sociomantic commented on relnotes/pkg.migration.md:
```diff
@@ -0,0 +1,19 @@
+* Packaging
+
+  * The `OPTS` and `ARGS` variables used when defining packages have become built-ins now, so they should be modified **in-place** (using `OPTS.update()`, `OPTS['xxx']`, and `ARGS.extend()`, `ARGS.append()` or any other functions that mutates the object).
```
Is there any reason to prefer `ARGS.extend` versus `ARGS.append`?  What's the expected difference?

leandro-lucarella-sociomantic replied:
Append in only for one element, extend will inject a list into the list.

So:

```py
[1, 2].append(3) # [1, 2, 3]
[1, 2].append([3, 4]) # [1, 2, [3, 4]]
[1, 2].extend([3, 4]) # [1, 2, 3, 4]
```

joseph-wakeling-sociomantic reviewed (approved):
This all LGTM (it reflects well my experience of updating 4 projects to use the release candidate).  I have some minor tweak suggestions, but these needn't be considered obligatory to fix.

leandro-lucarella-sociomantic commented:
Updated with fixes to joe's comments. @mihails-strasuns-sociomantic  / @gavin-norman-sociomantic please merge if you don't have any further comments!
